pkg/api/common/net.go

```diff
@@ -37,3 +41,17 @@ func IP4BroadcastAddress(n *net.IPNet) net.IP {
 	}
 	return last
 }
+
+// GetVNETSubnetIDComponents extract subscription, resourcegroup, vnetname, subnetname from the vnetSubnetID
+func GetVNETSubnetIDComponents(vnetSubnetID string) (string, string, string, string, error) {
+	vnetSubnetIDRegex := `^\/subscriptions\/([^\/]*)\/resourceGroups\/([^\/]*)\/providers\/Microsoft.Network\/virtualNetworks\/([^\/]*)\/subnets\/([^\/]*)$`
+	re, err := regexp.Compile(vnetSubnetIDRegex)
+	if err != nil {
+		return "", "", "", "", err
+	}
+	submatches := re.FindStringSubmatch(vnetSubnetID)
+	if len(submatches) != 5 {
+		return "", "", "", "", fmt.Errorf("unable to match regexp")
```

Comment:
changed from `return "", "", "", "", err` because it was returning a nil error and `if len(submatches) != 5` from `if len(submatches) != 4` because we expect 5 matches, not 4 (the first one is the entire string).

This was a bug in all apis except agentpoolonly v20180331 (the function was fixed in that one but not the others --> yay duplicate code) : validateVNET would never fail on invalid vnetID because it would find 5 submatches != 4 and exit with nil error.

Reply:
Good find!

Reply:
Curious: what is the fifth regex substring match that we're not doing anything with?

Reply:
Sorry, you answered the question above.

Unless that regex package is weird, we probably have a sub-optimal expression, as we should be able to formulate a regex result that includes only that matched patterns and not a throwaway match of the entire string as well.

Reply:
I tested on https://regex-golang.appspot.com/assets/html/index.html and got 4 matches... I wonder if it's the single quotes

Reply:
Hmm so this is because that's how the go regexp `FindStringSubmatch` function works  https://golang.org/pkg/regexp/#Regexp.FindStringSubmatch: `FindSubmatch returns a slice of slices holding the text of the leftmost match of the regular expression in b and the matches, if any, of its subexpressions, as defined by the 'Submatch' descriptions in the package comment.`